README
NAME
    Minecraft::SectionFilter - Strip/Process magical Â§ characters from
    minecraft

VERSION
    version 0.003003

SYNOPSIS
        use Minecraft::SectionFilter;
        while(<$some_stream_of_text>){
            if( $ENV{MODE} eq 'STRIP' ) {
                print(strip_sections($_))
            }
            else {
                print(ansi_encode_sections($_));
            }
        }

FUNCTIONS
  translate_sections
    Parse a string into a series of elements;

        my (@list) = translate_sections($string)

    Resulting list will be a list of hashrefs, either:

        { type => text , content => "the string itself" }

    or

        { type => section, section_code => $char }

  strip_sections
    Strip section codes from a string.

        my $output = strip_sections( $input );

  ansi_encode_sections
    Translate section codes to Term::ANSIColor color codes.

        STDOUT->print( ansi_encode_sections( $minecraft_string ) );

SEE ALSO
    Minecraft::RCON which has a similar feature, except its not
    user-acessible/reusable.

AUTHOR
    Kent Fredric <kentfredric@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Kent Fredric
    <kentfredric@gmail.com>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Release history for Minecraft-SectionFilter

0.003004 2017-03-10T09:07:40Z 8b08b37

0.003003 2017-03-10T08:53:47Z 50ff0ba
 - Bugfix: Avoid test failures with -Ddefault_inc_excludes_dot

 [Dependencies::Stats]
 - Dependencies changed since 0.003002, see misc/*.deps* for details
 - configure: +1 (recommends: ↑1)
 - develop: +14 ↑4 -3 (recommends: +1, suggests: ↑2)
 - runtime: -1
 - test: +2 (recommends: ↑1 ↓1)

0.003002 2014-08-15T23:11:19Z
 [00 Trivial]
 - CPANDAY!
 - no code changes.
 - reduce whitespace.

 [Dependencies::Stats]
 - Dependencies changed since 0.003001, see misc/*.deps* for details
 - develop: ↑1 (suggests: ↑1)

 [Misc]
 - Improve source side POD to reduce build side whitespace.

0.003001 2014-08-12T11:34:43Z
 [Bugfix]
 - require a Test::More without a broken subtest {}

 [Dependencies::Stats]
 - Dependencies changed since 0.003000, see misc/*.deps* for details
 - test: ↑1

0.003000 2014-08-12T01:28:34Z
 [00 Major]
 - Some changes that may affect hand-operated downstreams.

 [01 Toolkit]
 - Switched to EUMM
 - "Latest" dependencies now only recommendations
 - https://questhub.io/realm/perl/quest/5388cc12bbd0be5b2500011f

 [Dependencies::Stats]
 - Dependencies changed since 0.002000, see misc/*.deps* for details
 - build: -1
 - configure: +1 -1 (recommends: +1)
 - develop: +45 -2 (recommends: -1, suggests: +1 ↑1)
 - test: +2 ↓1 -2 (recommends: +3)

 [Misc]
 - moved to kentnl org
 - now legal META.yml
 - Source now UTF-8 instead of Latin1 (lol!)
 - Report Versions test uses Test::ReportPrereqs
 - Compile Test uses Test::Compile::PerFile

0.002000 2013-04-07T21:26:10Z
 [Dependencies::Stats]
 - Dependencies changed since 0.001002, see misc/*.deps* for details
 - runtime: +1

 [Feature]
 - Now warns when a un-assigned section code is seen in the ansi filter. 
 - If the section code is an upper case version of a lower case known one, lower case one is used and a warning is
   printed to STDERR
 - If the section code is not know, in either upper case or lower case, a warning is printed to STDERR and a
   placeholder inserted into the output in the form <unknow section O>

 [Packaging]
 - uses Module::Build 0.4004 test_requires

0.001002 2013-04-04T00:08:25Z
 [BugFix]
 - Depend on at least Term::ANSIColor 4 to avoid "bright_*" failing.

 [Dependencies::Stats]
 - Dependencies changed since 0.001001, see misc/*.deps* for details
 - build: ↑1
 - configure: ↑1
 - develop: (suggests: ↑1)
 - runtime: ↑1

 [Packaging]
 - Utilise Module::Build 0.4004 test_requires

0.001001 2013-03-30T18:59:02Z
 [Documentation]
 - Minor Layout issues
 - General synopsis for the function <ansi_encode_sections> was missing.

0.001000 2013-03-30T18:49:56Z
 - First version.


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Strip/Process magical \x{a7} characters from minecraft",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Minecraft-SectionFilter",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.010",
  "NAME" => "Minecraft::SectionFilter",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Sub::Exporter::Progressive" => 0,
    "Term::ANSIColor" => "4.00",
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => "0.9501",
    "utf8" => 0
  },
  "VERSION" => "0.003004",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Sub::Exporter::Progressive" => 0,
  "Term::ANSIColor" => "4.00",
  "Test::More" => "0.9501",
  "strict" => 0,
  "utf8" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/Minecraft/SectionFilter.pm
use 5.010;    # state
use strict;
use warnings;

package Minecraft::SectionFilter;

our $VERSION = '0.003004';

# ABSTRACT: Strip/Process magical § characters from minecraft

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Sub::Exporter::Progressive -setup => {
  exports => [qw( translate_sections strip_sections ansi_encode_sections )],
  groups  => {
    default => [qw( strip_sections ansi_encode_sections )],
  },
};

use Carp qw( carp );

















sub translate_sections {
  my ($line) = @_;

  state $section = chr 0xA7;

  my (@out);
  while ( length $line > 0 ) {
    if ( $line =~ /\A([^$section]+)/msx ) {
      push @out, { type => text =>, content => "$1", };
      substr $line, 0, length "$1", q{};
      next;
    }
    if ( $line =~ /\A$section(.)/msx ) {
      push @out, { type => section =>, section_code => "$1" };
      substr $line, 0, 2, q{};
    }

  }
  return @out;
}









sub _section_to_stripped {
  my ($section) = @_;
  return $section->{content} if 'text' eq $section->{type};
  return q{};
}

sub strip_sections {
  my ($section_string) = @_;
  return join q{}, map { _section_to_stripped($_) } translate_sections($section_string);
}

sub _ansi_translation_table {
  return state $translation_table = {
    0 => 'black',
    1 => 'blue',
    2 => 'green',
    3 => 'cyan',
    4 => 'red',
    5 => 'magenta',
    6 => 'yellow',
    7 => 'white',
    8 => 'bright_black',
    9 => 'bright_blue',
    a => 'bright_green',
    b => 'bright_cyan',
    c => 'bright_red',
    d => 'bright_magenta',
    e => 'bright_yellow',
    f => 'bright_white',

    l => 'bold',
    m => 'concealed',
    n => 'underscore',
    o => 'reverse',

    r => 'reset',
  };
}

sub _warn {
  my (@args) = @_;
  return carp( sprintf '[%s] %s', __PACKAGE__, join q{ }, @args );
}

sub _warnf {
  my (@args) = @_;
  my $format = '[%s] ' . shift;
  return carp( sprintf $format, __PACKAGE__, @args );
}

sub _section_to_ansi {
  my ($section) = @_;
  return $section->{content} unless 'section' eq $section->{type};
  state $colorize = do {
    require Term::ANSIColor;
    \&Term::ANSIColor::color;
  };
  state $trt = _ansi_translation_table();
  my ($code) = $section->{section_code};
  if ( exists $trt->{$code} ) {
    return $colorize->( $trt->{$code} );
  }
  if ( exists $trt->{ lc $code } ) {
    _warnf( 'uppercase section code "%s" (ord=%s)', $section->{section_code}, ord $section->{section_code} );
    return $colorize->( $trt->{ lc $code } );
  }
  _warnf( 'unknown section code "%s" (ord=%s)', $section->{section_code}, ord $section->{section_code} );
  return '<unknown section ' . $section->{section_code} . '>';
}









sub ansi_encode_sections {
  my ($section_string) = @_;
  return join q{}, map { _section_to_ansi($_) } translate_sections($section_string);
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Minecraft::SectionFilter - Strip/Process magical § characters from minecraft

=head1 VERSION

version 0.003004

=head1 SYNOPSIS

    use Minecraft::SectionFilter;
    while(<$some_stream_of_text>){
        if( $ENV{MODE} eq 'STRIP' ) {
            print(strip_sections($_))
        }
        else {
            print(ansi_encode_sections($_));
        }
    }

=head1 FUNCTIONS

=head2 translate_sections

Parse a string into a series of elements;

    my (@list) = translate_sections($string)

Resulting list will be a list of hashrefs, either:

    { type => text , content => "the string itself" }

or

    { type => section, section_code => $char }

=head2 strip_sections

Strip section codes from a string.

    my $output = strip_sections( $input );

=head2 ansi_encode_sections

Translate section codes to Term::ANSIColor color codes.

    STDOUT->print( ansi_encode_sections( $minecraft_string ) );

=head1 SEE ALSO

L<Minecraft::RCON|Minecraft::RCON> which has a similar feature, except its not user-acessible/reusable.

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.020.
use strict;
use warnings;

use Module::Build 0.28;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.4206"
  },
  "configure_requires" => {
    "Module::Build" => "0.4206"
  },
  "dist_abstract" => "Strip/Process magical \x{a7} characters from minecraft",
  "dist_author" => [
    "Kent Fredric <kentfredric\@gmail.com>"
  ],
  "dist_name" => "Minecraft-SectionFilter",
  "dist_version" => "0.002001",
  "license" => "perl",
  "module_name" => "Minecraft::SectionFilter",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "Sub::Exporter::Progressive" => 0,
    "Term::ANSIColor" => "4.00",
    "perl" => "v5.10.0",
    "strict" => 0,
    "utf8" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => "1.001003"
  }
);


my %fallback_build_requires = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Module::Build" => "0.4206",
  "Test::More" => "1.001003"
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;
